metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala

```diff
@@ -189,202 +152,237 @@ class ScalametaLanguageServer(
       renameProvider = true,
       codeActionProvider = true
     )
-    InitializeResult(capabilities)
+    Task(Right(InitializeResult(capabilities)))
   }
 
-  override def shutdown(): Task[ShutdownResult] = Task {
+  // TODO(olafur): make it easier to invoke fluid services from tests
+  def shutdown(): Unit = {
     logger.info("Shutting down...")
     cancelEffects.foreach(_.cancel())
-    connection.cancelAllActiveRequests()
-    ShutdownResult()
   }
 
-  private def onChangedFile(
-      path: AbsolutePath
-  )(fallback: AbsolutePath => Unit): Unit = {
-    val name = PathIO.extension(path.toNIO)
-    logger.info(s"File $path changed, extension=$name")
-    name match {
-      case "semanticdb" => fileSystemSemanticdbSubscriber.onNext(path)
-      case "compilerconfig" => compilerConfigSubscriber.onNext(path)
-      case _ => fallback(path)
+  val services: Services = Services.empty
+    .requestAsync[InitializeParams, InitializeResult]("initialize") { params =>
+      initialize(params)
     }
-  }
-
-  override def onChangeWatchedFiles(changes: Seq[FileEvent]): Unit =
-    changes.foreach {
-      case FileEvent(
-          Uri(path),
-          FileChangeType.Created | FileChangeType.Changed
-          ) =>
-        onChangedFile(path.toAbsolutePath) { _ =>
-          logger.warn(s"Unknown file extension for path $path")
+    .request[JsValue, JsValue]("shutdown") { _ =>
+      shutdown()
+      JsNull
+    }
+    .notification[JsValue]("exit") { _ =>
+      logger.info("exit(0)")
+      sys.exit(0)
+    }
+    .requestAsync[TextDocumentPositionParams, CompletionList](
```

Comment:
Yes, the former delegates, I made the distinction to reduce ceremony in adding new handlers. All handlers are currently synchronous except the codeActions handler that is truly asynchronous since it does a workspaceEdit roundtrip.